toy_2d/src/ilqr.py

```diff
@@ -0,0 +1,333 @@
+         [∂²l/∂u∂x, ∂²l/∂u²]], evaluated at xk, uk
+        """
+        H = np.zeros((self.nx + self.nu, self.nx + self.nu))
+
+        # TODO: Compute the hessian
+        H[:self.nx,:self.nx] = self.Q
+        H[self.nx:, self.nx:] = self.R
+
+        return H
+
+    def terminal_cost(self, xf: np.ndarray) -> float:
+        """
+        :param xf: state
+        :return: Lf(xf), the running cost incurred by xf
+        """
+        return 0.5*(xf - self.x_goal).T @ self.Qf @ (xf - self.x_goal)
+
+    def grad_terminal_cost(self, xf: np.ndarray) -> np.ndarray:
+        """
+        :param xf: final state
+        :return: ∂Lf/∂xf
+        """
+
+        grad = np.zeros((self.nx))
+
+        # TODO: Compute the gradient
+        x_grad = (xf- self.x_goal).T @ self.Qf
+        grad = x_grad
+        return grad
+        
+    def hess_terminal_cost(self, xf: np.ndarray) -> np.ndarray:
+        """
+        :param xf: final state
+        :return: ∂²Lf/∂xf²
+        """ 
+
+        H = np.zeros((self.nx, self.nx))
+        H = self.Qf
+
+        # TODO: Compute H
+
+        return H
+
+    def forward_pass(self, xx, uu, dd, KK, lr):
+        """
+        :param xx: list of states, should be length N
+        :param uu: list of inputs, should be length N-1
+        :param dd: list of "feed-forward" components of iLQR update, should be length N-1
+        :param KK: list of "Feedback" LQR gain components of iLQR update, should be length N-1
+        :return: A tuple (xtraj, utraj) containing the updated state and input
+                 trajectories after applying the iLQR forward pass
+        """
+
+        xtraj = [np.zeros((self.nx,))] * self.N
+        utraj = [np.zeros((self.nu,))] * (self.N - 1)
+        xtraj[0] = xx[0]
+
+        # TODO: compute forward pass
+        for i in range(self.N-1):
+            utraj[i] = uu[i] + KK[i]@(xtraj[i] - xx[i]) + lr*dd[i]
+            # if(utraj[i][0]<0 and utraj[i][1]>0): utraj[i] = np.array([0,0])
+            #TO DO
+            # xtraj[i+1] = quad_sim.F(xtraj[i], utraj[i], self.dt)
+            xtraj[i+1] = self.sim_forward(xtraj[i], utraj[i])
+        return xtraj, utraj
+
+    def get_linearized_discrete_dynamics(self, state, action):
+        q = np.array([state[0], state[2], state[4]])
+        v = np.array([state[1], state[3], state[5]])
+        self.lcs.set_linearization_point(q,v, action)
+        A, B_gen = self.lcs.get_A_matrix(), self.lcs.get_B_matrix()
+        pmat = self.lcs.get_P_matrix()
+        B = B_gen@pmat
+        C_mat = self.lcs.get_C_matrix()
+        G_mat = self.lcs.get_G_matrix()
+        J_mat = self.lcs.get_J_matrix()
+        H_mat = self.lcs.get_H_matrix()
+        #access original dynamics to get the original lambdas
+        curr_lam = self.get_lamda(state, action)
+        eps = 1e-3
+        # check if there is a contact at all. if not use the normal dynamics itself as cube is in air
+        if(np.argwhere(curr_lam>eps).shape[0]>0):
```

Comment:
Are you sure this is doing what you're expecting?  The vector `lambda` contains normal forces, tangential forces, _and_ slack variables which generally equal the relative contact velocity.  I'm not certain if no contact necessarily means those slack variables (the contact velocities) are zero.  Thus, it's possible that this `if` statement is satisfied more often than you're expecting (i.e. the polytope could be in the air, but the slack variable isn't zero, so you go into this `if` statement as if the polytope was contacting the ground).  With the structure of all the matrices, it might turn out that this is inconsequential (e.g. maybe if only one of the slack variables is non-zero but the forces are zero, then we get `A_d = 0` and `B_d = 0` such that we still recover the non-contact dynamics), though you may want to convince yourself if this is true or not.

Reply:
That's a really good point. I modified the code so that only the first 12 values (which contain the normal and the tangential forces) in the lambda vector are checked to be non-zero. If they are then modification is made to A and B matrices.